### ISSUE TITLE
Issues as first time user

### ISSUE THREAD

**Issue by patrickheeney** [jspm/jspm-cli#1524]
## **TLDR**: https://github.com/jspm/jspm-cli/issues/1524#issuecomment-183820079

I installed the beta of jspm and these are some problems I encountered as a new user.

## --Issue 1--

Running `jspm` outputs what appears to be a command: `jspm inspect npm:source-map`, but errors for me:

```
jspm inspect [--forks]           View all installed package versions
  inspect npm:source-map         View the versions and ranges of a package

$ jspm inspect npm:source-map

     Installed versions of npm:source-map

warn Package npm:source-map not found.
```

**EDIT:** After seeing this: `To inspect individual package constraints, use jspm inspect registry:name` I realize that registry is npm and name is the module name. I previously thought `npm:source-map` was a command to print the source of each npm module. Perhaps it could be: `jspm inspect [npm|github]:[module-name]` or something to denote that this isn't a static command.

## Issue 2

I tried to run `jspm install` but seem to have some issues with the registry. I was trying to track these down in `~/.jspm/jspm-cache` by experimenting with adding new entries to `registry.json` but it re-fetched from git on each run, overwriting my changes. Can there be a CLI command to disable cache updating: `jspm install --no-cache-fetch`?

## Issue 3

The previous error was after I tried to add an override. One of my dependencies seems to be using `engine.io-client` but it does not resolve correctly from npm: `jspm install npm:engine.io-client`, but seems to be ok from github: `jspm install github:socketio/engine.io-client`. The problem is I can't update my local cached registry as above, and I can't override the entire package. Perhaps you can support a syntax like the overrides section below. I don't want to declare it as a dependency because I don't know what is using it, what version, and my `jspm` section in `package.json` is dynamically generated (we don't use this dependency). 

```
$ jspm install npm:engine.io-client
     Updating registry cache...

err  Error on locate for github:component/global/archive/v2.0.1.tar.gz
GitHub packages must be of the form owner/repo.

err  Installing npm:engine.io-client@1.6.8, no version match for github:rase-/node-XMLHttpRequest@add%252Fssl-support

    "overrides": {
      "engine.io-client": "github:socketio/engine.io-client",
      "engine.io-client": {
          "resolve": "github:socketio/engine.io-client"
       }
    }
```

## Issue 4

The error messages don't really make sense to me. One of my dependencies triggered this error, but how do I fix it? I searched my codebase for hours looking for `github:rase-/node-XMLHttpRequest` with no result. The only place I found it was in the [registry](https://github.com/jspm/registry/blob/master/package-overrides/npm/engine.io-client%401.5.1.json#L3) which is for a different version than it is trying to resolve. It would be helpful to know which dependency has engine.io-client so I can debug this upstream?

```
$ jspm install npm:engine.io-client
     Updating registry cache...

err  Error on locate for github:component/global/archive/v2.0.1.tar.gz
GitHub packages must be of the form owner/repo.

err  Installing npm:engine.io-client@1.6.8, no version match for github:rase-/node-XMLHttpRequest@add%252Fssl-support

to

err  Error on locate for github:component/global/archive/v2.0.1.tar.gz from some-dependency@123.123 -> other-dependency@345
GitHub packages must be of the form owner/repo.

err  Installing npm:engine.io-client@1.6.8, no version match for github:rase-/node-XMLHttpRequest@add%252Fssl-support from some-dependency@123.123 -> other-dependency@345
```

**EDIT 1:** I renamed [this file](https://github.com/jspm/registry/blob/master/package-overrides/npm/engine.io-client%401.5.1.json) and it installed correctly. It may also be because I added `"engine.io-client": "github:socketio/engine.io-client"` to my registry cache. It seems that that it is applying the overrides even though the version does not match. Related: #758

**EDIT 2:** Perhaps there can be a `jspm install --log debug` that will output what is resolving to what, so I can figure out what is going on. 

## --Issue 5--

The error on this is also pretty confusing: `jspm install npm:webpack@1.12.13`. The [dependency](https://github.com/webpack/webpack/blob/v1.12.13/package.json#L15) for this package is: `"node-libs-browser": ">= 0.4.0 <=0.6.0"` which causes `jspm` to look for `npm:node-libs-browser@0.6.0`, [which does not exist](https://github.com/webpack/node-libs-browser/releases). The same command with npm directly: `npm install webpack@1.12.13` installs correctly to the right version:

```
$ jspm install npm:webpack@1.12.13

err  Installing npm:webpack@1.12.13, no version match for npm:node-libs-browser@0.6.0


$ npm install webpack@1.12.13

webpack@1.12.13 node_modules/webpack
├── node-libs-browser@0.5.3 (https-browserify@0.0.0, tty-browserify@0.0.0, constants-browserify@0.0.1, path-browserify@0.0.0, punycode@1.4.0, string_decoder@0.10.31, os-browserify@0.1.2, process@0.11.2, assert@1.3.0, domain-browser@1.1.7, querystring-es3@0.2.1, timers-browserify@1.4.2, events@1.1.0, stream-browserify@1.0.0, vm-browserify@0.0.4, readable-stream@1.1.13, util@0.10.3, console-browserify@1.1.0, url@0.10.3, http-browserify@1.7.0, buffer@3.6.0, browserify-zlib@0.1.4, crypto-browserify@3.2.8)
└── watchpack@0.2.9 (async@0.9.2, chokidar@1.4.2)
```

EDIT: Solved this with #759. Also going to see if this can be fixed upstream as `6.0` is not a valid version.

---

I have tested this on the latest stable, latest beta, and master with similar errors.

**Comment by patrickheeney** [comment]
This got a bit long, and I resolve a few, so going to try and TLDR this.

1. Confusion with the help screen. Nothing here.
2. Proposal for `--no-cache-fetch` to prevent registry from refreshing to aid in debugging. 
3. Proposal for overriding entire modules in `package.json`: `jspm install npm:engine.io-client` to pull from github package instead. 
4. Bug where `jspm` applies a package override `engine.io-client@1.5.1.json` to a package that is requesting `engine.io-client@1.6.8`. Also in here is proposal for debug output to aid in debugging.
5. Possible bug. Dependency of `"node-libs-browser": ">= 0.4.0 <=0.6.0"` looks for package version `0.6.0` which doesn't exist. Npm resolves correctly. Worked around with a PR.

**Comment by guybedford** [comment]
Sure, thanks its this sort of beginner feedback which is so useful to know where the catches are.


> 1. Confusion with the help screen. Proposal for improving wording.

Added https://github.com/jspm/jspm-cli/commit/6394529345d69dff2bc306bf6ce05210b6f7131c.

> 2. Proposal for --no-cache-fetch to prevent registry from refreshing to aid in debugging.

Manually editing the registry is not supported in this way, rather you can alter it via `jspm registry config jspm` and using a manual local clone. I'd rather not add an out-of-sync registry option. Rather I think this could be handled through a custom registry mirror today or future offline workflows. Created https://github.com/jspm/registry-endpoint/issues/12 to at least warn here.

> 3. Proposal for overriding entire modules in package.json: jspm install npm:engine.io-client to pull github:socketio/engine.io-client instead. Not dep overriding, but entire module.

Yes this will definitely be needed in future.

> 4. Bug where jspm applies a package override engine.io-client@1.5.1.json to a package that is requesting engine.io-client@1.6.8. Also in here is proposal for debug output to aid in debugging.

This is by design, overrides apply to all versions semver compatible with the override version. This way non-breaking upgrades to packages allow the existing local overrides to continue to apply, forming a smooth upgrade path.

> 5. Possible bug. Dependency of "node-libs-browser": ">= 0.4.0 <=0.6.0" looks for package version 0.6.0 which doesn't exist. Npm resolves correctly. Worked around with a PR.

It was by design that we don't support arbitrary version ranges. It's nice having simple version resolution rules, but as this becomes more and more impractical, we can look to putting implementation pressure on https://github.com/jspm/jspm-cli/issues/1253.

**Comment by patrickheeney** [comment]
Awesome, thank you! 

I think this is the only outstanding:

> This is by design, overrides apply to all versions semver compatible with the override version. This way non-breaking upgrades to packages allow the existing local overrides to continue to apply, forming a smooth upgrade path.

This makes sense, but I am not sure how to fix it. It seems at some point they [changed a dependency upstream](https://github.com/socketio/engine.io-client/blob/master/package.json#L28) and the [current override](https://github.com/jspm/registry/blob/master/package-overrides/npm/engine.io-client%401.5.1.json#L3) is incorrect. What would the override need to look like to essentially say: `drop dependency ["xmlhttprequest"] and add ["xmlhttprequest-ssl": "1.5.1"]?

**Comment by guybedford** [comment]
@patrickheeney if you send a PR to the registry that updates to the correct version for https://github.com/jspm/registry/blob/master/package-overrides/npm/engine.io-client%401.5.1.json on the version number it changed on, then that should fix the override?

**Comment by patrickheeney** [comment]
Not certain yet how this all works:

I assumed it would apply all engine.io-client.json files that matched the semver? Or does it only apply the first found and sort them by highest version?

If it only applies one, then it seems I would just need to create a blank .json file to prevent the other from loading. If it applies all matches, then how does it apply the dependencies overrides? Overriding all dependencies in favor of object, or does it simply replace the key => values and I would need to drop one and define another?

**Comment by guybedford** [comment]
@patrickheeney yes exactly it will only apply the most recent override from the registry, and blank overrides can be used to "disable" overrides after a certain version.
